ccl/changefeedccl: TestChangefeedRandomExpressions failed

ccl/changefeedccl.TestChangefeedRandomExpressions [failed](https://tanzanite.cluster.engflow.com/invocations/default/89948c0b-9482-4777-b565-cecd3c558783?testReportRun=4&testReportShard=12&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on release-24.2 @ [e07b00c604d971580a1300bc46a075ca48396e53](https://github.com/cockroachdb/cockroach/commits/e07b00c604d971580a1300bc46a075ca48396e53):


```
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES ((-1):::INT8, (-1):::INT8, NULL, (-0.09572368115186691):::FLOAT8, 0.3671309677636271:::FLOAT8, '4714-11-24 BC':::DATE, '2020-04-15 08:18:37.000186':::TIMESTAMP, '3000-01-01 00:00:00+00':::TIMESTAMPTZ, '-17 years -7 mons -124 days -12:43:40.22403':::INTERVAL, false, 7.892299539993146759E+35:::DECIMAL, 'X':::STRING, '\x22':::BYTES, 'ffffffff-ffff-ffff-ffff-ffffffffffff':::UUID, '6814:20bc:d6c4:ee0f:eb95:1080:f5fd:cdeb/82':::INET, '{"c": false}':::JSONB);
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES (15995:::INT8, 315403413:::INT8, 1:::INT8, (-0.0):::FLOAT8, 0.6287780390477413:::FLOAT8, '1981-09-22':::DATE, '4714-11-24 00:00:00 BC':::TIMESTAMP, '0001-01-01 00:00:00+00':::TIMESTAMPTZ, '11 years 3 mons 46 days 07:18:35.322345':::INTERVAL, NULL, (-13831222.72939908354):::DECIMAL, e'W:\x1bO\x05le\x11':::STRING, '\x00':::BYTES, '00000000-0000-0000-0000-000000000000':::UUID, '0.0.0.0/0':::INET, '1':::JSONB);
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES ((-29053):::INT8, (-1295317890):::INT8, 32767:::INT8, 0.22394979000091553:::FLOAT8, 0.760926326569046:::FLOAT8, '1971-03-22':::DATE, '2012-11-22 17:00:22.000793':::TIMESTAMP, '3000-01-01 00:00:00+00':::TIMESTAMPTZ, '1 mon 1 day':::INTERVAL, false, 218271.1710460755762:::DECIMAL, '':::STRING, '\xff':::BYTES, '00000000-0000-0000-0000-000000000000':::UUID, 'ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff':::INET, 'true':::JSONB);
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES (1:::INT8, (-1256183347):::INT8, 2056423856944306580:::INT8, '+Inf':::FLOAT8, 0.44868738114204487:::FLOAT8, '4714-11-24 BC':::DATE, '1986-11-28 10:02:03.00005':::TIMESTAMP, '2031-01-31 06:58:56.00051+00':::TIMESTAMPTZ, '00:00:00':::INTERVAL, false, 0.05435718528687338011:::DECIMAL, e'\'':::STRING, '\xff':::BYTES, 'f9e87b18-afd2-4ffc-93de-580adaf11bc3':::UUID, '5a6b:f332:1662:399b:b815:f959:4e9e:deeb/101':::INET, '{"b": "Ozg", "foo": null}':::JSONB);
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES ((-25025):::INT8, (-1487847063):::INT8, NULL, 0.7338573336601257:::FLOAT8, 1.2072446982084315:::FLOAT8, 'infinity':::DATE, '0001-01-01 00:00:00':::TIMESTAMP, '294276-12-31 23:59:59.999999+00':::TIMESTAMPTZ, '-47 years -8 mons -561 days -18:27:35.886909':::INTERVAL, NULL, (-911.1334488595475129):::DECIMAL, e'-\x068+5':::STRING, '\xff':::BYTES, 'e37e9d76-8404-4de3-96ba-71fdbb69c8b5':::UUID, '0.0.0.0/0':::INET, 'true':::JSONB);
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES (0:::INT8, 804009754:::INT8, 6343684396126218800:::INT8, 'NaN':::FLOAT8, 0.7656107997824307:::FLOAT8, '-infinity':::DATE, '2001-01-01 00:00:00 BC':::TIMESTAMP, '3000-01-01 00:00:00+00':::TIMESTAMPTZ, NULL, true, NULL, e'\U00002603':::STRING, '\xff':::BYTES, '9b9b24d0-5705-44d0-8ae8-e0d5716a2588':::UUID, '5a64:1693:43c0:2042:e00c:e85:3e64:4dc3/26':::INET, '["foo", null, 0.997149624413304]':::JSONB);
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES ((-32768):::INT8, (-1000435317):::INT8, 3480978684259436415:::INT8, 1.401298464324817e-45:::FLOAT8, (-1.0):::FLOAT8, '1970-01-01':::DATE, '1982-02-16 14:16:12.000626':::TIMESTAMP, '2021-11-13 17:51:06.00078+00':::TIMESTAMPTZ, '-29 years -2 mons -706 days -24:48:45.523998':::INTERVAL, true, (-4994.732132326854848):::DECIMAL, e'\'':::STRING, '\xcec11852ba':::BYTES, NULL, '0.0.0.0/0':::INET, '{"FYB": 0.580472915139266, "a": [{"foobar": 0.8089732119776573}, null], "aiTeX": ["foo", 1.4392804397730685], "vZ": null}':::JSONB);
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES ((-1):::INT8, (-127195124):::INT8, (-2541759397246338228):::INT8, (-0.43208423256874084):::FLOAT8, (-0.6513408575495625):::FLOAT8, '1984-03-26':::DATE, '1988-11-17 22:56:40.000776':::TIMESTAMP, '2009-02-16 18:45:29.000204+00':::TIMESTAMPTZ, '00:00:00':::INTERVAL, false, (-8212804441071846.684):::DECIMAL, e'\U00002603':::STRING, '\x00':::BYTES, NULL, 'b3eb:ead4:d156:ce5a:6d7f:88d8:4b1b:d73d/126':::INET, '["c"]':::JSONB);
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES (NULL, NULL, NULL, (-0.5794339776039124):::FLOAT8, 'NaN':::FLOAT8, '5874897-12-31':::DATE, '2028-04-19 05:09:17.000247':::TIMESTAMP, '2001-01-01 00:00:00+00 BC':::TIMESTAMPTZ, NULL, false, 'Infinity':::DECIMAL, '1E':::STRING, '\x5a0e48c5d20caf':::BYTES, 'ffffffff-ffff-ffff-ffff-ffffffffffff':::UUID, 'ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff':::INET, '{}':::JSONB);
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES (1:::INT8, 819881833:::INT8, 32767:::INT8, 2.225335121154785:::FLOAT8, 1.1751521316329239:::FLOAT8, '1986-09-12':::DATE, '1981-10-10 11:31:36.000646':::TIMESTAMP, NULL, '00:00:00':::INTERVAL, false, 21840938219.11073745:::DECIMAL, 'X':::STRING, '\x22':::BYTES, '5b1466ef-fbf9-407b-b1f0-c63de2bd02c4':::UUID, '0.0.0.0/0':::INET, '{"test": "json"}':::JSONB);
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES ((-11279):::INT8, 111686193:::INT8, 6536852776021058502:::INT8, 'NaN':::FLOAT8, (-1.3050796973347614):::FLOAT8, '5874897-12-31':::DATE, '1990-11-01 19:25:16.000281':::TIMESTAMP, NULL, '290 years':::INTERVAL, false, 2.354751949811069179E+29:::DECIMAL, e't\x17\fsc\x0ezS\x0e':::STRING, '\x0733375835a5772398':::BYTES, '3952059a-4267-4a01-b879-f1a1231d9976':::UUID, '97.240.209.121/11':::INET, '{"0wEWFL6RJ": [false], "bar": [false, false], "c": {"foobar": false}, "foo": 0.2654623341578454}':::JSONB);
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES (22678:::INT8, NULL, (-8779571282506485922):::INT8, 1.0:::FLOAT8, 0.2529776467862588:::FLOAT8, '5874897-12-31':::DATE, '0001-01-01 00:00:00':::TIMESTAMP, '2001-12-08 22:49:18.000585+00':::TIMESTAMPTZ, '-5 years -8 mons -319 days -15:37:13.130848':::INTERVAL, false, (-1.234E+401):::DECIMAL, e'R6\x174\x06\b|,':::STRING, '\xc8059f44fd':::BYTES, '41f61374-6b94-45ba-9715-bd393a04d70c':::UUID, '168.55.14.23/3':::INET, '{"test": "json"}':::JSONB);
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES ((-1):::INT8, 32767:::INT8, (-32768):::INT8, (-0.5827024579048157):::FLOAT8, 0.040240264596431374:::FLOAT8, '1976-12-01':::DATE, '3000-01-01 00:00:00':::TIMESTAMP, '3000-01-01 00:00:00+00':::TIMESTAMPTZ, '1 day':::INTERVAL, true, 2.465945105241191132E+34:::DECIMAL, e'~fU\x04|':::STRING, '\xf06127465e9f':::BYTES, 'fb9d378e-f070-493b-9813-e58fb4175ee9':::UUID, '0.0.0.0/0':::INET, '[[{"b": true, "eRb": 0.23771124091940268}, true, false], "foobar", "9mk"]':::JSONB);
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES ((-32768):::INT8, 1:::INT8, 8619968220600075605:::INT8, 0.0:::FLOAT8, 1.0:::FLOAT8, '1996-08-27':::DATE, '2007-04-22 10:23:07.000368':::TIMESTAMP, '0001-01-01 00:00:00+00':::TIMESTAMPTZ, '-2 years -11 mons -9 days -11:13:16.58293':::INTERVAL, NULL, 3.023366097043482927E+34:::DECIMAL, e'\nIgBB\x1f':::STRING, '\xe29883':::BYTES, 'ffffffff-ffff-ffff-ffff-ffffffffffff':::UUID, '0.0.0.0/0':::INET, '{"bar": [0.7212657925190564], "foobar": false, "oV5g8EG": false}':::JSONB);
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES (1972:::INT8, 196283471:::INT8, 7089341518779661299:::INT8, 1.401298464324817e-45:::FLOAT8, 0.09333997876252553:::FLOAT8, '1974-05-20':::DATE, '1978-01-09 22:20:58.000356':::TIMESTAMP, '2018-05-31 10:52:37.000903+00':::TIMESTAMPTZ, '1 mon 1 day':::INTERVAL, false, (-1.234E+401):::DECIMAL, '':::STRING, '\x39':::BYTES, 'e4b73a16-7243-4483-9dec-4f46ee71931f':::UUID, '2510:4f50:8095:d5e6:4eb4:2184:69a:dede/52':::INET, 'false':::JSONB);
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES ((-18661):::INT8, NULL, 1160922053682521402:::INT8, 0.23313841223716736:::FLOAT8, 0.6464779147235082:::FLOAT8, '-infinity':::DATE, NULL, '2021-09-03 15:31:01.000377+00':::TIMESTAMPTZ, NULL, true, 'Infinity':::DECIMAL, '':::STRING, NULL, 'ffffffff-ffff-ffff-ffff-ffffffffffff':::UUID, '0.0.0.0/0':::INET, '{"Nbqx": {"0Stpe": true}, "a": false, "c": [[0.8714049735392884], true, false], "foo": 1.5710120581464122}':::JSONB);
        INSERT INTO seed (_int2, _int4, _int8, _float4, _float8, _date, _timestamp, _timestamptz, _interval, _bool, _decimal, _string, _bytes, _uuid, _inet, _jsonb) VALUES (NULL, 32767:::INT8, NULL, 1.0:::FLOAT8, 0.6435059771570697:::FLOAT8, '-infinity':::DATE, '2001-01-01 00:00:00 BC':::TIMESTAMP, '0001-01-01 00:00:00+00':::TIMESTAMPTZ, '43 years 8 mons 866 days 15:35:05.754815':::INTERVAL, true, (-55902.67816933677789):::DECIMAL, e'G;\x11Cq0\x18':::STRING, '\xff':::BYTES, '00000000-0000-0000-0000-000000000000':::UUID, '0.0.0.0/0':::INET, '1':::JSONB);
        DELETE FROM seed WHERE rowid NOT IN (SELECT rowid FROM seed ORDER BY rowid LIMIT 100);
        ALTER TABLE seed ADD COLUMN _enum greeting;
        UPDATE seed SET _enum = enum_range('hi'::greeting)[rowid%7];
    changefeed_test.go:1153: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE e'\x07\x07a;.%]T':::STRING NOT SIMILAR TO e'\nt\x1fWX\x17\x06':::STRING
    changefeed_test.go:1169: Expecting statement CREATE CHANGEFEED WITH schema_change_policy='stop' AS SELECT rowid FROM seed WHERE e'\x07\x07a;.%]T':::STRING NOT SIMILAR TO e'\nt\x1fWX\x17\x06':::STRING to emit 100 events
    testfeed_test.go:301: creating external connection
    testfeed_test.go:304: ran create external connection
    changefeed_test.go:1153: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE seed._string NOT SIMILAR TO sha256(e'\\xe05098':::BYTES::BYTES)::STRING
    changefeed_test.go:1169: Expecting statement CREATE CHANGEFEED WITH schema_change_policy='stop' AS SELECT rowid FROM seed WHERE seed._string NOT SIMILAR TO sha256(e'\\xe05098':::BYTES::BYTES)::STRING to emit 87 events
    testfeed_test.go:301: creating external connection
    testfeed_test.go:304: ran create external connection
    changefeed_test.go:1153: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE true
    changefeed_test.go:1169: Expecting statement CREATE CHANGEFEED WITH schema_change_policy='stop' AS SELECT rowid FROM seed WHERE true to emit 100 events
    testfeed_test.go:301: creating external connection
    testfeed_test.go:304: ran create external connection
    changefeed_test.go:1153: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE false
    changefeed_test.go:1165: Skipping predicate false because it returned no rows
    changefeed_test.go:1153: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE seed._bool
    changefeed_test.go:1169: Expecting statement CREATE CHANGEFEED WITH schema_change_policy='stop' AS SELECT rowid FROM seed WHERE seed._bool to emit 50 events
    testfeed_test.go:301: creating external connection
    testfeed_test.go:304: ran create external connection
    changefeed_test.go:1153: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE NULL
    changefeed_test.go:1165: Skipping predicate NULL because it returned no rows
    changefeed_test.go:1153: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE seed._string LIKE seed._string
    changefeed_test.go:1169: Expecting statement CREATE CHANGEFEED WITH schema_change_policy='stop' AS SELECT rowid FROM seed WHERE seed._string LIKE seed._string to emit 86 events
    testfeed_test.go:301: creating external connection
    testfeed_test.go:304: ran create external connection
    changefeed_test.go:1153: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE (seed._bool AND (seed._string SIMILAR TO seed._string))
    changefeed_test.go:1169: Expecting statement CREATE CHANGEFEED WITH schema_change_policy='stop' AS SELECT rowid FROM seed WHERE (seed._bool AND (seed._string SIMILAR TO seed._string)) to emit 42 events
    testfeed_test.go:301: creating external connection
    testfeed_test.go:304: ran create external connection
    changefeed_test.go:1200: invalid regular expression: error parsing regexp: missing closing ]: ``[)$``
    --- FAIL: TestChangefeedRandomExpressions/kafka (3.36s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=4</code>
 - <code>shard=12</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedRandomExpressions.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-45761

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -1302,21 +1302,40 @@ func TestChangefeedRandomExpressions(t *testing.T) {
 			err = assertPayloadsBaseErr(context.Background(), seedFeed, assertedPayloads, false, false)
 			closeFeedIgnoreError(t, seedFeed)
 			if err != nil {
-				code := pgerror.GetPGCode(err)
 				// Skip errors that may come up during SQL execution. If the SQL query
 				// didn't fail with these errors, it's likely because the query was built in
 				// a way that did not have to execute on the row that caused the error, but
 				// the CDC query did.
-				switch code {
-				case pgcode.ConfigFile,
-					pgcode.DatetimeFieldOverflow,
-					pgcode.InvalidEscapeCharacter,
-					pgcode.InvalidEscapeSequence,
-					pgcode.InvalidParameterValue,
-					pgcode.InvalidRegularExpression:
-					t.Logf("Skipping statement %s because it encountered pgerror %s: %s", createStmt, code, err)
+				// Since we get the error that caused the changefeed job to
+				// fail from scraping the job status and creating a new
+				// error, we unfortunately don't have the pgcode and have to
+				// rely on known strings.
+				validPgErrs := []string{
+					"cannot subtract infinite dates",
+					"regexp compilation failed",
+					"invalid regular expression",
+					"error parsing GeoJSON",
+					"error parsing EWKT",
+					"geometry type is unsupported",
+					"should be of length",
+					"dwithin distance cannot be less than zero",
+					"parameter has to be of type Point",
+					"expected LineString",
+					"no locations to init GEOS",
+				}
+				containsKnownPgErr := func(e error) (interface{}, bool) {
+					for _, v := range validPgErrs {
+						if strings.Contains(e.Error(), v) {
+							return nil, true
+						}
+					}
+					return nil, false
+				}
+				if _, contains := errors.If(err, containsKnownPgErr); contains {
+					t.Logf("Skipping statement %s because it encountered pgerror %s", createStmt, err)
 					continue
 				}
+
 				t.Fatal(err)
 			}
 			numNonTrivialTestRuns++

